lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs

```diff
@@ -0,0 +1,164 @@
+    , ErrNetworkUnreachable (..)
+    , ErrPostTx (..)
+    , NetworkLayer (..)
+    )
+import Cardano.Wallet.Primitive.Types
+    ( Block, Tx (..), TxWitness (..) )
+import Control.Arrow
+    ( left )
+import Control.Exception
+    ( Exception )
+import Control.Monad.Catch
+    ( throwM )
+import Control.Monad.Trans.Except
+    ( ExceptT (..) )
+import Data.Proxy
+    ( Proxy (..) )
+import Network.HTTP.Client
+    ( Manager, defaultManagerSettings, newManager )
+import Network.HTTP.Types.Status
+    ( status404 )
+import Servant.API
+    ( (:<|>) (..) )
+import Servant.Client
+    ( BaseUrl (..), ClientM, Scheme (..), client, mkClientEnv, runClientM )
+import Servant.Client.Core
+    ( ServantError (..), responseStatusCode )
+import Servant.Links
+    ( Link, safeLink )
+
+-- | Constructs a network layer with the given @Jormungandr@ client.
+mkNetworkLayer :: Monad m => Jormungandr m -> NetworkLayer m
+mkNetworkLayer _httpBridge = NetworkLayer
+    { nextBlocks = undefined
+    , networkTip = undefined
+    , postTx = undefined
+    }
+
+-- | Creates a jormungandr 'NetworkLayer' using the given connection
+-- settings.
+newNetworkLayer
+    :: Int -> IO (NetworkLayer IO)
+newNetworkLayer port = mkNetworkLayer <$> newHttpBridge port
+  where
+    newHttpBridge = undefined
+
+{-------------------------------------------------------------------------------
+                            Jormungandr Client
+-------------------------------------------------------------------------------}
+
+-- | Endpoints of the jormungandr REST API.
+data Jormungandr m = HttpBridge
+    { getBlock
+        :: BlockId -> ExceptT ErrNetworkUnreachable m Block
+    , getTipId
+        :: ExceptT ErrNetworkTip m BlockId
+    , getDecendantIds
+        :: BlockId -> Int -> ExceptT ErrGetDecendants m [BlockId]
+    , postSignedTx
+        :: (Tx, [TxWitness])-> ExceptT ErrPostTx m ()
+    }
+
+-- | Error while trying to get decendants
+data ErrGetDecendants
+    = ErrGetDecendantsNetworkUnreachable ErrNetworkUnreachable
+    | ErrGetDecendantsParentNotFound
+    deriving (Show, Eq)
+
+instance Exception ErrGetDecendants
+
+-- | Construct a Jormungandr-client
+--
+-- >>> mgr <- newManager defaultManagerSettings
+-- >>> j = mkJormungandr mgr (BaseUrl Http "localhost" 8080 "")
+-- >>> runExceptT $ getTipId j
+-- Right (BlockId (Hash {getHash = "26c640a3de09b74398c14ca0a137ec78"}))
+mkJormungandr
+    :: Manager -> BaseUrl -> Jormungandr IO
+mkJormungandr mgr baseUrl = HttpBridge
+    { getBlock = \hash -> ExceptT $ do
+        let ctx = safeLink api (Proxy @GetBlock) hash
+        run (cGetBlock hash) >>= defaultHandler ctx
+
+    , getTipId = ExceptT $ do
+        run cGetTipId >>= \case
+            Left (FailureResponse e) | responseStatusCode e == status404 ->
+              return $ Left ErrNetworkTipNotFound
+            x -> do
+                let ctx = safeLink api (Proxy @GetTipId)
+                left ErrNetworkTipNetworkUnreachable <$> defaultHandler ctx x
```

Comment:
I'm kind-of not comfortable yet with how the `servant-client` requests are written, the error handling or the tests, but I'm not sure its worth focusing on, since we already have the same thing going on for the `HttpBridge`.

Reply:
What do you mean by "focusing on" ?

Reply:
Making an attempt at refactoring.

Not in the pr now, but for instance I find the duplication of `parentId (Just count)` off
```haskell
        run (cGetBlockDescendantIds parentId (Just count))  >>= \case
            Left (FailureResponse e) | responseStatusCode e == status404 ->
              return $ Left ErrGetDescendantsParentNotFound
            x -> do
                let ctx = safeLink
                        api
                        (Proxy @GetBlockDescendantIds)
                        parentId
                        (Just count)
                left ErrGetDescendantsNetworkUnreachable <$> defaultHandler ctx x
```